core/src/main/c/share/dedup.cpp

```diff
@@ -0,0 +1,647 @@
+    return a.hi > b.hi || a.lo > b.lo;
+}
+
+inline bool operator<(const int256 &a, const int256 &b) {
+    return a.hi < b.hi || a.lo < b.lo;
+}
+
+template<typename LambdaDiff>
+inline int64_t branch_free_search(const index_t *array, int64_t count, int64_t value_index, LambdaDiff compare) {
+    const index_t *base = array;
+    int64_t n = count;
+    while (n > 1) {
+        int64_t half = n / 2;
+        MM_PREFETCH_T0(base + half / 2);
+        MM_PREFETCH_T0(base + half + half / 2);
+        auto diff = compare(value_index, base[half].i);
+        base = (diff > 0) ? base + half : base;
+        n -= half;
+    }
+    if (compare(value_index, base[0].i) == 0) {
+        return base - array;
+    }
+    if (base - array + 1 < count && compare(value_index, base[1].i) == 0) {
+        return base - array + 1;
+    }
+    return -1;
+}
+
+template<typename LambdaDiff>
+int64_t merge_dedup_long_index_int_keys(
+        const uint64_t *src,
+        int64_t src_lo,
+        const int64_t src_hi_incl,
+        const index_t *index,
+        int64_t index_lo,
+        const int64_t index_hi_incl,
+        index_t *dest_index,
+        const LambdaDiff compare
+) {
+    int64_t &src_pos = src_lo;
+    int64_t &index_pos = index_lo;
+    index_t *dest = dest_index;
+
+    bit_vector used_indexes = {};
+    while (src_pos <= src_hi_incl && index_pos <= index_hi_incl) {
+        // Perform normal merge until the timestamp matches.
+        if (src[src_pos] < index[index_pos].ts) {
+            (*dest).ts = src[src_pos];
+            // The first bit is used by merge procedures to know where to take the value from. 1 indicates column, 0 O3 data
+            (*dest).i = src_pos | (1ull << 63);
+            dest++;
+            src_pos++;
+        } else if (src[src_pos] > index[index_pos].ts) {
+            *dest++ = index[index_pos];
+            index_pos++;
+        } else {
+            // index_ts == src_ts
+            const uint64_t conflict_ts = src[src_pos];
+            const index_t *conflict_index_start = &index[index_pos];
+
+            // Find end of the conflict in index
+            const auto conflict_end_pos = index_pos +
+                                          branch_free_search_lower<index_t>(&index[index_pos],
+                                                                            index_hi_incl - index_pos + 1,
+                                                                            {conflict_ts + 1, 0});
+
+            // binary search for matching index record for every source record
+            const int64_t binary_search_len = conflict_end_pos - index_pos;
+
+            // track all found index records
+            used_indexes.reset(binary_search_len);
+            while (src_pos <= src_hi_incl && src[src_pos] == conflict_ts) {
+                (*dest).ts = conflict_ts;
+                const int64_t matched_index = branch_free_search(conflict_index_start, binary_search_len,
```

Comment:
nit: would be great to check that we always pass properly sorted (sorted by ts + all dedup cols) index here as otherwise the binary search will break. Of course, that should be something only enabled in tests. WDYT?

Reply:
In future (definitely not in this PR), we could add a self-diagnosing config prop that would turn on costly extra checks in the database, such as this one.